spring-boot-project/spring-boot-tools/spring-boot-loader-tools/src/main/java/org/springframework/boot/loader/tools/Repackager.java

```diff
@@ -181,7 +182,7 @@ public void repackage(File destination, Libraries libraries,
 		}
 		destination = destination.getAbsoluteFile();
 		File workingSource = this.source;
-		if (alreadyRepackaged() && this.source.equals(destination)) {
+		if (alreadyRepackaged()) {
```

Comment:
I think it's a bit more complicated than that. What that check was doing before your change was checking a case where a user tries to repackage an archive that was already repackaged with a replacement of the source jar.

If the source and the destination are different and the source is already repackaged, I think there's something to be done (like adding a warning to indicate we can't repackage as the source is already a fat jar.

Reply:
The check `&& this.source.equals(destination)` was added to fix #15034
I looked at the pom.xml provided in the sample app : 

```
<plugin>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-maven-plugin</artifactId>
	<version>2.1.0.RELEASE</version><!-- change to 2.0.6.RELEASE -->
	<executions>
		<execution>
			<id>build-linux-executable</id>
			<goals>
			        <goal>repackage</goal>
			</goals>
			<configuration>
				<classifier>exec-linux</classifier>
				<executable>true</executable>
				<attach>false</attach>
			</configuration>
		</execution>  
		<execution>
			<id>build-repackage-executable</id>
			<goals>
				<goal>repackage</goal>
			</goals>
			<configuration>
				<executable>false</executable>
				<attach>true</attach>
			</configuration>
		</execution>
	</executions>
</plugin>
```
and from the error message 

`[ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin:2.1.0.RELEASE:repackage (build-repackage-executable) on project demo: Execution build-repackage-executable of goal org.springframework.boot:spring-boot-maven-plugin:2.1.0.RELEASE:repackage failed: Source must refer to an existing file, got C:\Users\issam\Downloads\demo\target\gear-exec-linux.jar.original`

I suspected that `build-repackage-executable` execution was running before `build-linux-executable`  execution ends.

Then I decided to give it a try and tested it with the provided sample